### PR TITLE
WIP: Read Saml tokens from XmlReader

### PR DESCRIPTION
**Pull request by AndersAbel**
Work in progress to solve #1042 , just opening this to get an early possibility for feedback.

Looks like this is going to work. Quite a few overloads were added. And I will need to extract a string representation (canonicalized) of the token during the reading, as it is required for TokenValidationParameters.IssuerSigningKeyResolver and SaveSigninToken.

### CODE REVIEW

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.cs [hunk 1]
```diff
@@ -261,6 +297,38 @@ protected virtual void ValidateTokenReplay(DateTime? expirationTime, string secu
...
+
+            if (validationParameters == null)
+                throw LogArgumentNullException(nameof(validationParameters));
+
+            if (validationParameters.SignatureValidator != null)
+                throw new NotImplementedException();
+
+            var samlToken = ReadSaml2Token(tokenReader);
+
+            return ValidateSignature(null, samlToken, validationParameters);
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.cs]: Why pass null, this changes the model for Saml2 token validation.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.cs [hunk 2]
```diff
@@ -201,17 +201,53 @@ public override ClaimsPrincipal ValidateToken(string token, TokenValidationParam
                 throw LogExceptionMessage(new ArgumentException(FormatInvariant(TokenLogMessages.IDX10209, token.Length, MaximumTokenSizeInBytes)));
 
             var samlToken = ValidateSignature(token, validationParameters);
+
+            ValidateToken(token, validationParameters, out validatedToken, samlToken, out ClaimsIdentity identity);
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.cs]: why are two out params needed?

In general we need the same model for all token handling.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.cs [hunk 3]
```diff
@@ -261,6 +297,38 @@ protected virtual void ValidateTokenReplay(DateTime? expirationTime, string secu
...
+        /// <para>If the 'token' signature is validated, then the <see cref="Saml2SecurityToken.SigningKey"/> will be set to the key that signed the 'token'. It is the responsibility of <see cref="TokenValidationParameters.SignatureValidator"/> to set the <see cref="Saml2SecurityToken.SigningKey"/></para></remarks>
+        protected virtual Saml2SecurityToken ValidateSignature(XmlReader tokenReader, TokenValidationParameters validationParameters)
+        {
+            if (tokenReader == null)
+                throw LogArgumentNullException(nameof(tokenReader));
+
+            if (validationParameters == null)
+                throw LogArgumentNullException(nameof(validationParameters));
+
+            if (validationParameters.SignatureValidator != null)
```
[the remark this answers is not in the document]
Comment (AndersAbel) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.cs]: Yes, the SAML2 API that reads off an XmlReader would be different. The existing API accepting a string is still kept. But supporting the SignatureValidator when reading off an XmlReader would require the token to be written out to a string, just to later read it back from a string. Which IMHO wouldn't make sense.

**Comment by AndersAbel** [comment]
This is getting far too complex, so I'm closing this.

I'm now looking into adding the features I need by subclassing the Saml2SecurityTokenHandler instead. I will probably get back with a PR that splits the ValidateToken method into two parts, to allow me to call the second part after doing the first part in my own override.

**Comment by brentschmaltz** [comment]
@AndersAbel I am working on getting in the Saml2SecurityTokenHandler to accept an xmlreader.
One of the issues I am solving is that if the Saml token inside a WSFedMessage has a "/r/n", the signature will fail.
Essentially, our current code is broken. We need changes to our WsFederation message.
It's actually broken back to the originally release in WIF in 2008.

We we have this change in place, it should be natural for a Samlp serializer, similar to SamlSerializer.

**Pull request closed without merging**